File: CSharp/BrainHatSharp/GpioPinManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static WiringPiWrapper.WiringPiProxy;
using WiringPiWrapper;

namespace BrainHatSharp
{
    public static class GpioPinManager
    {
        //  Define pins used

        //  Hard Hat
        static int PinRightRising = 37;
        static int PinRightFalling = 38;
        static int PinRightBlink = 36;

        static int PinLeftRising = 7;
        static int PinLeftFalling = 11;
        static int PinLeftBlink = 13;


        static int PinLightStringA5 = 35;
        static int PinLightStringA4 = 32;
        static int PinLightStringA3 = 31;
        static int PinLightStringA2 = 29;
        static int PinLightStringA1 = 33;

        static int PinLightStringB5 = 23;
        static int PinLightStringB4 = 21;
        static int PinLightStringB3 = 24;
        static int PinLightStringB2 = 22;
        static int PinLightStringB1 = 26;

        static int PinHapticMotor = 12;

        static int PinPowerSwitch = 40;



        //  Proto board
        //static int PinRightRising = 40;
        //static int PinRightFalling = 38;
        //static int PinRightBlink = 36;

        //static int PinLeftRising = 7;
        //static int PinLeftFalling = 11;
        //static int PinLeftBlink = 13;



        //static int PinLightStringA5 = 29;
        //static int PinLightStringA4 = 31;
        //static int PinLightStringA3 = 32;
        //static int PinLightStringA2 = 33;
        //static int PinLightStringA1 = 35;

        //static int PinLightStringB5 = 26;
        //static int PinLightStringB4 = 23;
        //static int PinLightStringB3 = 24;
        //static int PinLightStringB2 = 21;
        //static int PinLightStringB1 = 19;

        //static int PinHapticMotor = 16;

        //  List of all pins in use (used for all off)
        static List<int> PinsInUse;

        //  Light strings
        public static LightString LightStringMaster;
        static LightString LightStringLeft;
        static LightString LightStringRight;


        /// <summary>
        /// Setup wiringPi and initialize our pins for output
        /// </summary>
        public static async Task SetupGpio()
        {

            WiringPiSetupPhys();
        
            PinsInUse = new List<int>();
            PinsInUse.Add(PinLeftRising);
            PinsInUse.Add(PinLeftFalling);
            PinsInUse.Add(PinLeftBlink);

            PinsInUse.Add(PinRightRising);
            PinsInUse.Add(PinRightFalling);
            PinsInUse.Add(PinRightBlink);

            PinsInUse.Add(PinLightStringA1);
            PinsInUse.Add(PinLightStringA2);
            PinsInUse.Add(PinLightStringA3);
            PinsInUse.Add(PinLightStringA4);
            PinsInUse.Add(PinLightStringA5);
            
            PinsInUse.Add(PinLightStringB1);
            PinsInUse.Add(PinLightStringB2);
            PinsInUse.Add(PinLightStringB3);
            PinsInUse.Add(PinLightStringB4);
            PinsInUse.Add(PinLightStringB5);

            PinsInUse.Add(PinHapticMotor);
            PinsInUse.Add(PinPowerSwitch);

            foreach (var nextPin in PinsInUse)
                PinMode(nextPin, WiringPiPinMode.Output);

            AllOff();

            foreach (var nextPin in PinsInUse)
            {
                if (nextPin == PinPowerSwitch)
                    continue;

                DigitalWrite(nextPin, WiringPiPinValue.High);
                await Task.Delay(333);
                DigitalWrite(nextPin, WiringPiPinValue.Low);
            }

            LightStringRight = new LightString(new int[] { PinLightStringB1, PinLightStringB2, PinLightStringB3, PinLightStringB4, PinLightStringB5 });
            LightStringLeft = new LightString(new int[] { PinLightStringA1, PinLightStringA2, PinLightStringA3, PinLightStringA4, PinLightStringA5 }, LightStringRight);
            LightStringMaster = LightStringLeft;

            //  test lights
            await LightStringMaster.StartFlashAsync(111, 333, 3);
            await Task.Delay(3000);
            await LightStringMaster.StartSequenceAsync(111, 333, false);
            await Task.Delay(5000);
            await LightStringMaster.Stop();
            await Task.Delay(1000);
            await LightStringMaster.StartSequenceAsync(111, 333, true);
            await Task.Delay(5000);
            await LightStringMaster.Stop();
            await Task.Delay(1000);

            LightStringLeft.SetPins(new int[] { 1, 0, 1, 0, 1 });
            LightStringRight.SetPins(new int[] { 0, 1, 0, 1, 0 });
            await Task.Delay(2000);

            BuzzMotor(0, 2000, 1);
            await Task.Delay(1000);
            BuzzMotor(111, 222, 3);

            await LightStringMaster.StartSequenceAsync(111, 333, true);
        }



        public static void AllOff()
        {
            foreach (var nextPin in PinsInUse)
                DigitalWrite(nextPin, WiringPiPinValue.Low);
        }


        public static async void CommandMode(bool on)
        {
            if (on)
            {
                BuzzMotor(333, 111, 3);
                await LightStringMaster.StartFlashAsync(111, 333, 3);
            }
            else
            {
                BuzzMotor(333, 0, 1);
                await LightStringMaster.StartSequenceAsync(111, 333, true);
            }
        }


        public static async void BuzzMotor(int intervalBetween, int duration, int times)
        {
            for(int i = 0; i < times; i++)
            {
                if (i > 0)
                    await Task.Delay(intervalBetween);
                DigitalWrite(PinHapticMotor, WiringPiPinValue.High);
                await Task.Delay(duration);
                DigitalWrite(PinHapticMotor, WiringPiPinValue.Low);
            }
        }

        public static void CommandTrigger2()
        {

            BuzzMotor(333, 111, 2);
        }

        public static async void LightLeftRising()
        {
            DigitalWrite(PinLeftRising, WiringPiPinValue.High);
            await Task.Delay(333);
            DigitalWrite(PinLeftRising, WiringPiPinValue.Low);
        }

        public static async void LightLeftFalling()
        {
            DigitalWrite(PinLeftFalling, WiringPiPinValue.High);
            await Task.Delay(222);
            DigitalWrite(PinLeftFalling, WiringPiPinValue.Low);
        }

        public static async void LightLeftBlink()
        {
            DigitalWrite(PinLeftFalling, WiringPiPinValue.High);
            DigitalWrite(PinLeftBlink, WiringPiPinValue.High);
            await Task.Delay(333);
            DigitalWrite(PinLeftFalling, WiringPiPinValue.Low);
            DigitalWrite(PinLeftBlink, WiringPiPinValue.Low);
        }

        public static async void LightRightRising()
        {
            DigitalWrite(PinRightRising, WiringPiPinValue.High);
            await Task.Delay(333);
            DigitalWrite(PinRightRising, WiringPiPinValue.Low);
        }

        public static async void LightRightFalling()
        {
            DigitalWrite(PinRightFalling, WiringPiPinValue.High);
            await Task.Delay(222);
            DigitalWrite(PinRightFalling, WiringPiPinValue.Low);
        }

        public static async void LightRightBlink()
        {
            DigitalWrite(PinRightFalling, WiringPiPinValue.High);
            DigitalWrite(PinRightBlink, WiringPiPinValue.High);
            await Task.Delay(333);
            DigitalWrite(PinRightFalling, WiringPiPinValue.Low);
            DigitalWrite(PinRightBlink, WiringPiPinValue.Low);
        }

        public static void EnableHapticMotor(bool enable)
        {
            if (enable)
                DigitalWrite(PinHapticMotor, WiringPiPinValue.High);
            else
                DigitalWrite(PinHapticMotor, WiringPiPinValue.Low);

        }

        public static void BrainBoardPower(bool on)
        {
            DigitalWrite(PinPowerSwitch, on ? WiringPiPinValue.High : WiringPiPinValue.Low);
        }




    }
}

File: brainHatServer/brainHat/GpioPinManager.cpp
#include <unistd.h>
#include "GpioPinManager.h"
#include "wiringPi.h"

using namespace std;

//  Set all pin numbers to zero
//  by default this will do nothing to the GPIO
//  this is overridden by host name mapping for certain devices below
//
int PinRightRising = 0;
int PinRightFalling = 0;
int PinRightBlink = 0;

int PinLeftRising = 0;
int PinLeftFalling = 0;
int PinLeftBlink = 0;

int PinLightStringA5 = 0;
int PinLightStringA4 = 0;
int PinLightStringA3 = 0;
int PinLightStringA2 = 0;
int PinLightStringA1 = 0;

int PinLightStringB5 = 0;
int PinLightStringB4 = 0;
int PinLightStringB3 = 0;
int PinLightStringB2 = 0;
int PinLightStringB1 = 0;

int PinHapticMotor = 0;

int PinPowerSwitch = 0;

//  Wrap wiringPi functions to do nothing with pin 0
void DigitalWrite(int pin, int value)
{
	if (pin != 0)
		digitalWrite(pin, value);
}
//
void AllPinsOff(vector<int> pins)
{
	for (auto it = pins.begin(); it != pins.end(); ++it)
	{
		if (*it != 0)
			DigitalWrite(*it, LOW);
	}
}
//
void AllPinsOn(vector<int> pins)
{
	for (auto it = pins.begin(); it != pins.end(); ++it)
	{
		if (*it != 0)
			DigitalWrite(*it, HIGH);
	}
}


// Setup GPIO
// this will set pin numbers for known host names where the prototypes have wired pins
//
void GpioManager::SetupGpio()
{
	if (HostName.compare("brainHelmet") == 0)
	{
		PinRightRising = 37;
		PinRightFalling = 38;
		PinRightBlink = 36;

		PinLeftRising = 0;
		PinLeftFalling = 0;
		PinLeftBlink = 0;

		PinLightStringA5 = 35;
		PinLightStringA4 = 32;
		PinLightStringA3 = 31;
		PinLightStringA2 = 29;
		PinLightStringA1 = 33;

		PinLightStringB5 = 23;
		PinLightStringB4 = 21;
		PinLightStringB3 = 24;
		PinLightStringB2 = 22;
		PinLightStringB1 = 26;

		PinHapticMotor = 12;

		PinPowerSwitch = 40;
	}
	else if (HostName.compare("brainHat") == 0)
	{
		PinLeftRising = 40;
		PinLeftFalling = 38;
		PinLeftBlink = 36;

		PinRightRising = 7;
		PinRightFalling = 11;
		PinRightBlink = 13;

		PinLightStringA5 = 29;
		PinLightStringA4 = 31;
		PinLightStringA3 = 32;
		PinLightStringA2 = 33;
		PinLightStringA1 = 35;

		PinLightStringB5 = 26;
		PinLightStringB4 = 23;
		PinLightStringB3 = 24;
		PinLightStringB2 = 21;
		PinLightStringB1 = 19;

		PinHapticMotor = 16;
		
		PinPowerSwitch = 0;
	}
	
	wiringPiSetupPhys();
	
	PinsInUse.push_back(PinLeftRising);
	PinsInUse.push_back(PinLeftFalling);
	PinsInUse.push_back(PinLeftBlink);
	//
	BlinkLightsLeft.push_back(PinLeftRising);
	BlinkLightsLeft.push_back(PinLeftFalling);
	BlinkLightsLeft.push_back(PinLeftBlink);
	BlinkLightsLeft.push_back(PinLeftFalling);
	BlinkLightsLeft.push_back(PinLeftRising);
	
	
	PinsInUse.push_back(PinRightRising);
	PinsInUse.push_back(PinRightFalling);
	PinsInUse.push_back(PinRightBlink);
	//
	BlinkLightsRight.push_back(PinRightRising);
	BlinkLightsRight.push_back(PinRightFalling);
	BlinkLightsRight.push_back(PinRightBlink);
	BlinkLightsRight.push_back(PinRightFalling);
	BlinkLightsRight.push_back(PinRightRising);

	PinsInUse.push_back(PinLightStringA1);
	PinsInUse.push_back(PinLightStringA2);
	PinsInUse.push_back(PinLightStringA3);
	PinsInUse.push_back(PinLightStringA4);
	PinsInUse.push_back(PinLightStringA5);
	LightStringLeft.push_back(PinLightStringA1);
	LightStringLeft.push_back(PinLightStringA2);
	LightStringLeft.push_back(PinLightStringA3);
	LightStringLeft.push_back(PinLightStringA4);
	LightStringLeft.push_back(PinLightStringA5);
            
	PinsInUse.push_back(PinLightStringB1);
	PinsInUse.push_back(PinLightStringB2);
	PinsInUse.push_back(PinLightStringB3);
	PinsInUse.push_back(PinLightStringB4);
	PinsInUse.push_back(PinLightStringB5);
	LightStringRight.push_back(PinLightStringB1);
	LightStringRight.push_back(PinLightStringB2);
	LightStringRight.push_back(PinLightStringB3);
	LightStringRight.push_back(PinLightStringB4);
	LightStringRight.push_back(PinLightStringB5);

	PinsInUse.push_back(PinHapticMotor);
	PinsInUse.push_back(PinPowerSwitch);
	
	for (auto it = PinsInUse.begin(); it != PinsInUse.end(); ++it)
	{
		if (*it != 0 )
			pinMode(*it, OUTPUT);
	}
	
	for (auto it = PinsInUse.begin(); it != PinsInUse.end(); ++it)
	{
		if (*it == PinPowerSwitch)
			continue;
		
		DigitalWrite(*it, HIGH);
	}
	DigitalWrite(PinHapticMotor, LOW);
}

GpioManager::GpioManager()
{
	
}

GpioManager::~GpioManager()
{
	
}


void GpioManager::StartThreadForHost()
{
	char host[1024];
	gethostname(host, 1024);
	
	HostName = string(host);
	
	//  only do something if this is a device with pins hooked up
	if(HostName.compare("brainHat") == 0 || HostName.compare("brainHelmet") == 0)
	{
		SetupGpio();
	
		Thread::Start();
	}
}

void GpioManager::Cancel()
{
	Thread::Cancel();
	AllOff();
}

void GpioManager::AllOn()
{
	AllPinsOn(PinsInUse);
}

void GpioManager::AllOff()
{
	AllPinsOff(PinsInUse);
}

void GpioManager::PowerToBoard(bool enable)
{
	DigitalWrite(PinPowerSwitch, enable ? HIGH : LOW);
}


void GpioManager::RunFunction()
{
	while (ThreadRunning)
	{
		switch (Mode)
		{
		case LightsOff:
			AllPinsOff(LightStringLeft);
			AllPinsOff(LightStringRight);
			AllPinsOff(BlinkLightsLeft);
			AllPinsOff(BlinkLightsRight);
			Sleep(250);
			break;
			
		case LightsOn:
			AllPinsOn(LightStringLeft);
			AllPinsOn(LightStringRight);
			AllPinsOn(BlinkLightsLeft);
			AllPinsOn(BlinkLightsRight);
			Sleep(250);
			break;
			
		case LightsFlash:
			RunLightsFlash();
			break;
			
		case LightsSequence:
			RunLightsSequence();
			break;
		}
	}
}



void GpioManager::RunLightsFlash()
{
	
	while (RunningLightsFlash())
	{
		Sleep(111);
		AllPinsOn(LightStringLeft);
		AllPinsOn(LightStringRight);
		AllPinsOn(BlinkLightsLeft);
		AllPinsOn(BlinkLightsRight);
		if (! RunningLightsFlash())
			return;
		Sleep(333);
		AllPinsOff(LightStringLeft);
		AllPinsOff(LightStringRight);
		AllPinsOff(BlinkLightsLeft);
		AllPinsOff(BlinkLightsRight);
	}
}



void GpioManager::RunLightsSequence()
{
	while (RunningLightsSequence())
	{
		for (int i = 0; i < LightStringLeft.size(); i++)
		{
			if (Mode != LightsSequence)
				return;
			Sleep(111);
			DigitalWrite(LightStringLeft[i], HIGH);
			DigitalWrite(LightStringRight[i], HIGH);
			DigitalWrite(BlinkLightsLeft[i], HIGH);
			DigitalWrite(BlinkLightsRight[i], HIGH);
			if (!RunningLightsSequence())
				return;
			Sleep(333);
			DigitalWrite(LightStringLeft[i], LOW);
			DigitalWrite(LightStringRight[i], LOW);
			DigitalWrite(BlinkLightsLeft[i], LOW);
			DigitalWrite(BlinkLightsRight[i], LOW);
		}
		Sleep(111);
		for (int i = LightStringLeft.size() - 1; i >= 0; i--)
		{
			if (!RunningLightsSequence())
				return;
			DigitalWrite(LightStringLeft[i], HIGH);
			DigitalWrite(LightStringRight[i], HIGH);
			DigitalWrite(BlinkLightsLeft[i], HIGH);
			DigitalWrite(BlinkLightsRight[i], HIGH);
			if (!RunningLightsSequence())
				return;
			Sleep(333);
			DigitalWrite(LightStringLeft[i], LOW);
			DigitalWrite(LightStringRight[i], LOW);	
			DigitalWrite(BlinkLightsLeft[i], LOW);
			DigitalWrite(BlinkLightsRight[i], LOW);
		}
	}
}